### PR TITLE
Enhancement: State Injection using Container and Injector Pattern

### LINKED ISSUE #440: State injection containers
@matt-gadd commented on [Thu Feb 16 2017](https://github.com/dojo/meta/issues/146)

We're now at a point where the widgeting system is nearly complete and for building components the property system suffices for passing of state and callbacks.

In larger apps this is a cumbersome and coupled way to get state and callbacks/actions down the tree of widgets when working with an application state store.

In react-redux the concept of a`Container` act as a kind of virtual widget which allow you to side load application state and actions. We have had similar concepts in the past with `externalState` that do similar but not quite as complete in functionality.

As part of the issue we should probably look at whether there is a place for an `actions` like mechanism (dojo/actions) or whether simple functions suffice.

### PR DESCRIPTION
**Pull request by agubler**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**

Implement the base classes required to support injecting context using the concept of an "Injector" to define the context and a "Container" to wrap a Widget and provide the bindings between the context and the component.

Resolves #440

### CODE REVIEW

File: src/Injector.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+export class BaseInjector<C> extends WidgetBase<InjectorProperties> {
+	public toInject(): C {
+		return <C> {};
+	}
+}
+
+/**
+ * Mixin that extends the supplied Injector class with the proxy `render` and passing the provded to `context` to the Injector
```
Comment (dylans) [inline src/Injector.ts]: provded -> provided

File: src/mixins/Container.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+		return <P> {};
+	},
+	getChildren<C>(inject: C, children: DNode[]): DNode[] {
+		return [];
+	}
+};
+
+/**
+ * Given the registered name of an Injector entry with property and child binding mappers, the
+ * container proxies the provided Widget and modifying the properties and children with the
```
Comment (dylans) [inline src/mixins/Container.ts]: modifying -> modifies

File: src/mixins/Container.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ */
+export interface Mappers {
+	getProperties: GetProperties;
+	getChildren: GetChildren;
+}
+
+/**
+ * Default noop Mappers for the container.
+ */
+const defautMappers: Mappers = {
```
Comment (dylans) [inline src/mixins/Container.ts]: defautMappers -> defaultMappers

File: src/mixins/Container.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * Given the registered name of an Injector entry with property and child binding mappers, the
+ * container proxies the provided Widget and modifying the properties and children with the
+ * instructions provided by the mappers using the context provideded by the registered Injector.
+ */
+export function Container<P extends WidgetProperties, T extends Constructor<WidgetBase<P>>>(
+	Base: T,
+	name: string,
+	{ getProperties = defautMappers.getProperties, getChildren = defautMappers.getChildren }: any = defautMappers
```
Comment (dylans) [inline src/mixins/Container.ts]: defautMappers -> defaultMappers

**Comment by agubler** [comment]
Thanks @dylan; I'd picked up on the default mappers locally 😀 Hadn't re-reviewed the rest yet.

**Comment by agubler** [comment]
The missing coverage is on the transpiled supers else block that can never be exercised.

**Comment by agubler** [comment]
@dylans addressed feedback.

**Comment by agubler** [comment]
I don't know what to do about the CI failures with Microsoft Edge here - The tests themselves all pass but the job hangs at the end.

**Pull request merged**